File: hackerrank/problem solving/algorithms/1 - warmup/10 - time conversion/time_conversion.js
// https://www.hackerrank.com/challenges/time-conversion/problem?isFullScreen=true

function main() {
  const time = readLine();
  console.log(to24HourTime(time));
}

// T: O(1)
// S: O(1) extra space
function to24HourTime(hour12) {
  let hour24 = hour12.substring(0, 8);
  const hour = hour12.substring(0, 2);
  const dayPeriod = hour12.substring(8, 9);

  if (
    (hour === '12' && dayPeriod === 'A') ||
    (hour !== '12' && dayPeriod === 'P')
  ) {
    const newHour = genNewHour(dayPeriod, hour);
    hour24 = hour24.replace(hour, newHour);
  }

  return hour24;
}

function genNewHour(dayPeriod, hour) {
  if (dayPeriod === 'A') return '00';
  let newHour = +hour;
  newHour += 12;
  return newHour.toString();
}

//////////////////////////////////////////////////

const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
const inputLines = [];
rl.on('line', (line) => inputLines.push(line));
rl.on('close', main);

function readLine() {
  return inputLines.shift();
}

File: hackerrank/problem solving/algorithms/1 - warmup/10 - time conversion/time_conversion.py
# https://www.hackerrank.com/challenges/time-conversion/problem?isFullScreen=true

def main():
    time = input()
    print(to_24_hour_time(time))


# T: O(1)
# S: O(1) extra space
def to_24_hour_time(hour_12):
    hour_24 = hour_12[:8]
    hour = hour_12[:2]
    day_period = hour_12[8:9]

    if (hour == '12' and day_period == 'A') or (hour != '12' and day_period == 'P'):
        new_hour = gen_new_hour(day_period, hour)
        hour_24 = hour_24.replace(hour, new_hour, 1)

    return hour_24


def gen_new_hour(day_period, hour):
    if day_period == 'A':
        return '00'
    new_hour = int(hour)
    new_hour += 12
    return str(new_hour)


if __name__ == '__main__':
    main()
